core/src/main/java/org/mule/util/store/InMemoryObjectStore.java

```diff
@@ -139,44 +139,75 @@ public void clear() throws ObjectStoreException
         }
     }
 
-    @Override
-    public void expire()
+    protected int expireAndCount()
     {
-        // this is not guaranteed to be precise, but we don't mind
-        int currentSize = store.size();
+        // first we trim the store according to max size
+        int expiredEntries = 0;
 
-        // first trim to maxSize if necessary
-        currentSize = trimToMaxSize(currentSize);
+        final long now = System.nanoTime();
 
-        // expire further if entry TTLs are enabled
-        if ((entryTTL > 0) && (currentSize != 0))
+        Map.Entry<?, ?> oldestEntry;
+
+        purge:
+        while ((oldestEntry = store.firstEntry()) != null)
         {
-            final long now = System.nanoTime();
-            int expiredEntries = 0;
-            Map.Entry<?, ?> oldestEntry;
+            Long oldestKey = (Long) oldestEntry.getKey();
+            long oldestKeyValue = oldestKey.longValue();
 
-            purge:
-            while ((oldestEntry = store.firstEntry()) != null)
+            if (TimeUnit.NANOSECONDS.toMillis(now - oldestKeyValue) >= entryTTL)
             {
-                Long oldestKey = (Long) oldestEntry.getKey();
-                long oldestKeyValue = oldestKey.longValue();
-
-                if (TimeUnit.NANOSECONDS.toMillis(now - oldestKeyValue) >= entryTTL)
-                {
-                    store.remove(oldestKey);
-                    expiredEntries++;
-                }
-                else
-                {
-                    break purge;
-                }
+                store.remove(oldestKey);
+                expiredEntries++;
             }
-
-            if (logger.isDebugEnabled())
+            else
             {
-                logger.debug("Expired " + expiredEntries + " old entries");
+                break purge;
             }
         }
+
+        return expiredEntries;
+    }
+
+    protected boolean isTrimNeeded(int currentSize)
+    {
+        return currentSize > maxEntries;
+    }
+
+    protected boolean isExpirationNeeded()
+    {
+        // this is not guaranteed to be precise, but we don't mind
+        int currentSize = store.size();
+
+        // should expire further if entry TTLs are enabled
+        return entryTTL > 0 && currentSize != 0;
+    }
+
+    @Override
+    public void expire()
+    {
+        int expiredEntries = doTrimAndExpire();
+
+        if (logger.isDebugEnabled() && expiredEntries != 0)
```

Comment:
why the expiredEntries != 0 condition?. Should log it anyway so if we put this class in debug mode we can actually know if it's not expiring entries.